### ISSUE TITLE
Dartanalyzer should report an error when catchError returns a different type

### ISSUE THREAD

**Issue by tomyeh** [dart-lang/sdk#35545]
Dartanalyzer doesn't detect any error/warning for the following code. However, there will be a run-time error: type 'String' is not a subtype of type 'FutureOr<List<dynamic>>'

It is hard to pick up this kind of errors by eyes.

> IMO, it is better to allow `Future.catchError()` to return another type like `Future.then<R>()` does. But, it is OK as long as dartanlyser can pick up the errors for us.

```
import "dart:async";

void main() {
  foo()
  .catchError((ex) {
  	return "error";
  });
}
Future<List> foo() async {
  throw "error";
}
```

Dart  SDK: 2.1

**Comment by stereotype441** [comment]
The reason the analyzer isn't catching this case is that the type of `Future<T>.catchError`'s argument is `Function`, which is too general to allow the type system to detect the problem; it allows any function whatsoever is allowed to be passed in, regardless of what it returns.

In an ideal world, we would add union types to the language (see https://github.com/dart-lang/language/issues/145), and then `catchError`'s argument could be changed to have a more precise type: `((dynamic) -> FutureOr<T>) | ((dynamic, StackTrace) -> FutureOr<T>)`.  Then the analyzer would detect the mistake in the code above.

But fully general support for union types isn't on the language funnel right now, so in the meantime we might want to could consider special-case logic in the analyzer to handle this case.  Accordingly I'm classifying this as an enhancement rather than a bug.

**Comment by srawlins** [comment]
I'm working on #35825, which I think is more-or-less a feature which solves this issue.

**Comment by srawlins** [comment]
Fixed in work for #35825
